fish-rust/src/builtins/string.rs

```diff
@@ -0,0 +1,493 @@
+        }
+
+        if streams.stdin_is_directly_redirected() && args.len() > optind {
+            string_error!(streams, BUILTIN_ERR_TOO_MANY_ARGUMENTS, args[0]);
+            return STATUS_INVALID_ARGS;
+        }
+
+        return self.handle(parser, streams, &mut optind, args);
+    }
+}
+
+/// This covers failing argument/option parsing
+enum StringError {
+    InvalidArgs(&'static wstr),
+    /// Like above but automatically `Invalid <opt_name> value '<arg>'`
+    InvalidValue,
+    NotANumber,
+    UnknownOption,
+}
+
+enum RegexError {
+    Compile(WString, pcre2::Error),
+    InvalidCaptureGroupName(WString),
+    InvalidEscape(WString),
+}
+
+impl RegexError {
+    fn print_error(&self, args: &[&wstr], streams: &mut io_streams_t) {
+        let cmd = args[0];
+        use RegexError::*;
+        match self {
+            Compile(pattern, e) => {
+                string_error!(
+                    streams,
+                    "%ls: Regular expression compile error: %ls\n",
+                    cmd,
+                    &WString::from(e.error_message())
+                );
+                string_error!(streams, "%ls: %ls\n", cmd, pattern);
+                string_error!(streams, "%ls: %*ls\n", cmd, e.offset().unwrap(), "^");
+            }
+            InvalidCaptureGroupName(name) => {
+                streams.err.append(wgettext_fmt!(
+                    "Modification of read-only variable \"%ls\" is not allowed\n",
+                    name
+                ));
+            }
+            InvalidEscape(pattern) => {
+                string_error!(
+                    streams,
+                    "%ls: Invalid escape sequence in pattern \"%ls\"\n",
+                    cmd,
+                    pattern
+                );
+            }
+        }
+    }
+}
+
+impl From<crate::wutil::wcstoi::Error> for StringError {
+    fn from(_: crate::wutil::wcstoi::Error) -> Self {
+        StringError::NotANumber
+    }
+}
+
+impl From<std::num::TryFromIntError> for StringError {
+    fn from(_: std::num::TryFromIntError) -> Self {
+        // the result of `try_into` failing on the result of fish_wcsto{i,l}
+        StringError::InvalidValue
+    }
+}
+
+impl From<&'static wstr> for StringError {
+    fn from(s: &'static wstr) -> Self {
+        StringError::InvalidArgs(s)
+    }
+}
+
+impl StringError {
+    fn print_error(
+        &self,
+        args: &[&wstr],
+        parser: &mut parser_t,
+        streams: &mut io_streams_t,
+        optarg: Option<&wstr>,
+        optind: usize,
+        option_name: &wstr,
+    ) {
+        let cmd = args[0];
+        use StringError::*;
+        match self {
+            InvalidArgs(_) | InvalidValue => {
+                let msg = match *self {
+                    InvalidArgs(s) => s.to_owned(),
+                    _ => sprintf!("Invalid %s value", option_name),
+                };
+                let error_msg = wgettext_fmt!("%ls: %ls '%ls'\n", cmd, msg, optarg.unwrap());
```

Comment:
> This translates only `"%ls: %ls '%ls'\n"`, not the dynamic messages.
We should rather put the already translated message inside `InvalidArgs`.
In practice it will probably work fine to store the translated string as `&'static wstr` in `InvalidArgs`, and format it here,
however that doesn't feel safe so I think we should store a WString that is already formatted.

So `wgettext()` at the "throw"-site, and replace this with `sprintf`?

Reply:
> So wgettext() at the "throw"-site, and replace this with sprintf?

either that or maybe get rid of the InvalidArgs and other error codes altogether.
They serve no use other than being printed, so we can use a WString just as well.
To avoid duplication we can define macros like `invalid_args!`

Reply:
Does this not work? I think this is closer to the original `string_error` occurrences.

```suggestion
                let template = wgettext_str(&sprintf!("string %%ls: %ls '%%ls'\n", msg));
                let error_msg = sprintf!(template, cmd, optarg.unwrap());
```

I find it kinda weird that `sprintf` just _works_ with a runtime-supplied template, probably explains why it is so slow

Reply:
> They serve no use other than being printed, so we can use a WString just as well.
To avoid duplication we can define macros like invalid_args!

I mean if we were gonna be fully idiomatic this would use [anyhow](https://docs.rs/anyhow/latest/anyhow/) and possibly [thiserror](https://docs.rs/thiserror/latest/thiserror/), but those operate with the [`Error`](https://doc.rust-lang.org/std/error/trait.Error.html)-trait which uses UTF8-strs.

Reply:
wgettext might also work with non-constant strings but we don't want that because it would make it hard for the translation tool to extract the set of strings to translate

Reply:
in C++, translation strings are enclosed in `_()` and gettext finds those when generating *.po files.
I assume we eventually want to have the same in Rust with the `wgettext` and `wgettext_fmt` instead of `_`

Reply:
I'm going to replace the printing of `InvalidArgs` to `string_error!(streams, "%ls: %ls\n", cmd, msg)`, so we can keep the same printing behaviour by "translating" `%ls: %ls\n` to `%ls : %ls\n`, but I don't know how that's supposed to be added

Reply:
yes French has different punctuation conventions.
I think translating `%ls: %ls\n` to `%ls : %ls\n` would work fine in practice although it's too clever for my taste.
Although this PR is maybe not the best place for this type of change.
Ideally it doesn't touch translation strings at all. Would it be easy to get the program name from the context?

Reply:
> Would it be easy to get the program name from the context?

You would need to send the program name/`args` to the `parse_opt`-function, which I find a very weird mix of error-handling and translation. You are translating two different things: 1. the punctuation convention and 2. the error-message, which just messes with abstractions and locks us into a very particular way of implementing the option/argument parsing, as you cannot just have this relatively pure function:

```rust
fn parse_opt(&mut self, optarg: Option<&wstr>, c: char) -> Result<(), StringError>
```

Since you need to know the name of the subcommand, we need to either (a) store a reference to it in every option struct, (b) have a method that reverse-engineers the name from e.g. `is_split0`, or (c) send the full `args`/`args[0]` to every invocation.

This gets messier for `FieldParseError`. Including the invalid field in the error message is completely fine, since you have that available anyways, but now you need to match and supply the context to the error message in `parse_opt`, instead of cleanly implementing `From<FieldParseError> for StringError`, and utilising the implicit `Into` of `?`.

With this context I don't think translating `%ls: %ls\n` to `%ls : %ls\n` even is particularly clever — we are translating two different things, and mixing it is limiting how we can implement this parsing.

https://github.com/fish-shell/fish-shell/pull/9854/files#diff-852e60f7ea0397cc63b03078dfc9c28a23092fa3d782cbb126400e048444e35aR69

I will fix this in a way that does not require changing translation strings, but with a `TODO` of fixing this once xgettext or whatever we are gonna use is available for Rust